File: tui/src/scala/tui/internal/breakableForeach.scala
package tui
package internal

object breakableForeach {
  sealed trait Res
  case object Continue extends Res
  case object Break extends Res

  implicit class BreakableForeachIterator[T](private val it: Iterator[T]) extends AnyVal {
    @inline
    def breakableForeach(f: T => Res): Unit = {
      var continue = true
      while (it.hasNext && continue)
        f(it.next()) match {
          case Continue =>
            ()
          case Break =>
            continue = false
        }
    }
  }
  implicit class BreakableForeachArray[T](private val ts: Array[T]) extends AnyVal {
    @inline
    def breakableForeach(f: (T, Int) => Res): Unit = {
      var continue = true
      var i = 0
      while (i < ts.length && continue) {
        f(ts(i), i) match {
          case Continue =>
            ()
          case Break =>
            continue = false
        }
        i += 1
      }
    }
  }
}

File: tui/src/java/tui/internal/BreakableForeach.java
package tui.internal;

import java.util.Iterator;
import java.util.function.BiFunction;
import java.util.function.Function;

public final class BreakableForeach {
  private BreakableForeach() {}

  public enum Res {
    Continue,
    Break
  }

  public static <T> void run(Iterator<T> it, Function<T, Res> f) {
    boolean cont = true;
    while (it.hasNext() && cont) {
      switch (f.apply(it.next())) {
        case Continue -> {}
        case Break -> cont = false;
      }
    }
  }

  public static <T> void run(T[] ts, BiFunction<T, Integer, Res> f) {
    boolean cont = true;
    int i = 0;
    while (i < ts.length && cont) {
      switch (f.apply(ts[i], i)) {
        case Continue -> {}
        case Break -> cont = false;
      }
      i += 1;
    }
  }
}
